library/core/src/iter/adapters/flatten.rs

```diff
@@ -159,6 +164,48 @@ where
 {
 }
 
+#[unstable(issue = "none", feature = "inplace_iteration")]
+unsafe impl<I, U, F> InPlaceIterable for FlatMap<I, U, F>
+where
+    I: InPlaceIterable,
+    U: KnownExpansionFactor + IntoIterator,
+{
+    const EXPAND_BY: Option<NonZeroUsize> = const {
+        match (I::EXPAND_BY, U::FACTOR) {
+            (Some(m), Some(n)) => m.checked_mul(n),
+            _ => None,
+        }
+    };
+    const MERGE_BY: Option<NonZeroUsize> = I::MERGE_BY;
+}
+
+#[unstable(issue = "none", feature = "inplace_iteration")]
+unsafe impl<I, U, F> SourceIter for FlatMap<I, U, F>
+where
+    I: SourceIter + TrustedFused,
+    U: IntoIterator,
+{
+    type Source = I::Source;
+
+    #[inline]
+    unsafe fn as_inner(&mut self) -> &mut I::Source {
+        // SAFETY: unsafe function forwarding to unsafe function with the same requirements
+        unsafe { SourceIter::as_inner(&mut self.inner.iter) }
+    }
+}
+
+#[rustc_specialization_trait]
+trait KnownExpansionFactor {
+    const FACTOR: Option<NonZeroUsize> = NonZeroUsize::new(1);
+}
+
+impl<T> KnownExpansionFactor for Option<T> {}
+impl<T> KnownExpansionFactor for option::IntoIter<T> {}
+impl<T> KnownExpansionFactor for Once<T> {}
+impl<T, const N: usize> KnownExpansionFactor for [T; N] {
+    const FACTOR: Option<NonZeroUsize> = NonZeroUsize::new(N);
+}
```

Comment:
Not sure if you're trying to be complete, but I think this could add: `option::{Iter, IterMut}`, `Result`, `result::{IntoIter, Iter, IterMut}`, `OnceWith`, and `array::IntoIter`.

And then adapters... or would that run into specialization woes?

Reply:
This is used for in-place iteration which means items will be moved through the pipeline so it's unlikely that items will be by-ref iterators. It's not impossible, e.g. when borrowing from some global state or something, but less likely.

Adapters also wouldn't make much sense for `flatten()`. But it could work for `flat_map()`. So that's an interesting idea.

Reply:
Done. It's not exhaustive but hopefully covers common shapes.